### PR TITLE
`hvncat`: Stronger argument checks

### PR DESCRIPTION
**Pull request by BioTurboNick**
Breaking down #41143 into smaller pieces.

This adds stronger argument checks.

### CODE REVIEW

**Review by simeonschaub — APPROVED** [review]
Would be nice to have a full fix for #41047 of course, but this seems like a reasonable fix for now.

**Comment by BioTurboNick** [comment]
The one failing tester is unrelated.

I recommend extra care be taken when merging these pieces; it's possible that one could inadvertently revert parts of another because there is some overlap.

File: base/abstractarray.jl [hunk 1]
```diff
@@ -2157,11 +2169,21 @@ function hvncat_fill!(A::Array, row_first::Bool, xs::Tuple)
 end
 
 _typed_hvncat(T::Type, dim::Int, ::Bool, xs...) = _typed_hvncat(T, Val(dim), xs...) # catches from _hvncat type promoters
-_typed_hvncat(::Type{T}, ::Val) where T = Vector{T}()
 _typed_hvncat(T::Type, ::Val{N}, xs::Number...) where N = _typed_hvncat(T, (ntuple(x -> 1, N - 1)..., length(xs)), false, xs...)
+
+function _typed_hvncat(::Type{T}, ::Val{N}) where {T, N}
+    N < 0 &&
```
Comment (JeffBezanson) [inline base/abstractarray.jl:2175]: Doesn't it need to be at least 1?

Comment (JeffBezanson) [reply base/abstractarray.jl:2175]: Oh I see, it's for 0-d arrays? I'm not sure that makes sense, since there is no dimension 0.

Comment (BioTurboNick) [reply base/abstractarray.jl:2175]: See #41197

**Pull request merged**